vortex-expr/src/lib.rs

```diff
@@ -1,27 +1,71 @@
 #![feature(iter_intersperse)]
 
+use std::any::Any;
+use std::collections::HashSet;
+use std::fmt::Debug;
 use std::sync::Arc;
 
+mod binary;
+mod column;
 pub mod datafusion;
-mod expr;
+mod identity;
+mod literal;
 mod operators;
+mod select;
 
-pub use expr::*;
+pub use binary::*;
+pub use column::*;
+pub use identity::*;
+pub use literal::*;
 pub use operators::*;
+pub use select::*;
+use vortex::Array;
+use vortex_dtype::field::Field;
+use vortex_error::{VortexExpect, VortexResult};
 
+/// [`VortexExpr`] represents logical operation on [`Array`]s
+pub trait VortexExpr: Debug + Send + Sync + PartialEq<dyn Any> {
+    fn as_any(&self) -> &dyn Any;
+
+    fn evaluate(&self, batch: &Array) -> VortexResult<Array>;
+
+    fn references(&self) -> HashSet<Field>;
+}
+
+/// Split expression into conjunctive normal form
 pub fn split_conjunction(expr: &Arc<dyn VortexExpr>) -> Vec<Arc<dyn VortexExpr>> {
+    split_inner(expr, vec![])
+}
+
+fn split_inner(
+    expr: &Arc<dyn VortexExpr>,
+    mut exprs: Vec<Arc<dyn VortexExpr>>,
+) -> Vec<Arc<dyn VortexExpr>> {
     match expr.as_any().downcast_ref::<BinaryExpr>() {
         Some(bexp) if bexp.op() == Operator::And => {
-            let mut exprs = split_conjunction(bexp.lhs());
-            exprs.extend_from_slice(&split_conjunction(bexp.rhs()));
+            let split = split_inner(bexp.lhs(), exprs);
```

Comment:
why do you need `let split` instead of just doing:

```
split_inner(bexp.lhs(), exprs);
split_inner(bexp.lhs(), exprs);
exprs
```

or even
```
split_inner(bexp.lhs(), exprs);
split_inner(bexp.lhs(), exprs)
```

Reply:
eh, I can change the signature of this, right now what you suggest won't work since it's all owned so it consumes the value and returns it